contributor-docs/adrs/adr-008-experimental-components.md

```diff
@@ -0,0 +1,70 @@
+# ADR 008: Lowering the barrier to entry for experimental components
+
+## Status
+
+Proposed
+
+&nbsp;
+
+## Context
+
+#### Recap: Drafts
+
+As we work on new components (or rewrites of old components), we often need to start with a "prototype"[^1] and not include them in the semantically versioned (semver-ed) main bundle.
+
+The approach that has served us well since Dec 2021 has been "drafts" (along with causing a lot of confusion).
+
+[ADR-006](./adr-006-drafts.md) explains our use of drafts, but here are the relevant parts:
+
+- When we are not sure about the API of a component, we put it in a different bundle: "primer/react/drafts"
+- A draft has experimental API, but follows the development conventions of primer/react
+- Components in drafts are accompanied by a design spec and go through the process of accessibility reviews. A component can check multiple boxes in the component lifecycle.
+- The contract with consumers is that you are opting into a component in which we don't have 100% confidence yet.
+
+[^1]: Calling this a prototype and not experimental, because that already means something else in the maturity lifecycle.
+
+&nbsp;
+
+#### Problems with drafts
+
+1. **Incompatible with semver**: Because the expectation with drafts is that we are not 100% confident about the API or structure of a component, we slip in breaking changes to draft components as minor releases to primer/react. On the consumer side, this makes every upgrade, even a patch, a potentially breaking upgrade. So, you are effectively opting out of your semver.
+
+1. **Confusion about maturity**: Because of the generic name, "drafts" causes a lot of confusion, especially with "experimental". It is a prototype or experiment, but NOT in the way the [component lifecycle]([https://primer.style/contribute/component-lifecycle]) defines "Experimental" maturity.
+
+1. **Need for experimental candidates**: As more teams build React components and look to upstream reusable components, it is natural to look at drafts as the place to put them. However, the barrier to entry into drafts isn't low (because we still follow design reviews, dev conventions, API reviews, etc.)
+
+Upstreaming a component that was built in the context of a product requires re-evaluating the baked-in assumptions and most likely a change in the API to make it useful for other contexts.
+
+&nbsp;
+
+## Decision
+
+1. "Experimental" components, as defined by the component lifecycle, should not be part of semantically versioned npm package for `@primer/react`.
+
+2. Each experimental component should be independently versioned with semver. Breaking changes should be tagged as major versions.
+
+3. To support different conventions and processes for experimental components, we should keep them in a new repository `primer/react-candidates`
```

Comment:
> Could we publish a @primer/react-candidates (or whatever we want to call it) from the primer/react repo?

Sorry, some nitpicking first 😅

The key is that 
1) experimental components are not part of the npm package `@primer/react`
2) each component is independently versioned like `@primer/react-commentbox`, not in a combined package like `@primer/react-candidates`

We can publish these components independently while keeping them in primer/react repository (repo, not npm package) or from another repository like primer/react-candidates or both depending on the component.

I personally think it's harder to maintain different dev conventions/practices and code ownership from a single repository, but I can go either way on this.

Reply:
I do prefer using the `candidates` name instead of `shared`, because it contributes to the idea that it's not fully adopted within Primer. 👍 

Am I right in thinking that we don't want a `@primer/react/candidates` in order to mitigate the versioning issue? I do agree with @colebemis  that it'd be easier to publish candidates with the main package if not.

Reply:
> I do prefer using the candidates name instead of shared, because it contributes to the idea that it's not fully adopted within Primer. 👍

💯 

> Am I right in thinking that we don't want a @primer/react/candidates in order to mitigate the versioning issue?

Yep.

I think I haven't done a good job at distinguishing the strategy for npm package and code repository. 

Right now, we have 1 code repository [github.com/primer/react](http://github.com/primer/react) and 1 npm package: [`@primer/react`](https://www.npmjs.com/@primer/react).

For the npm package, the proposal is publish experimental components outside of the npm package for `@primer/react` to keep the confidence/trust that there are no breaking changes in minor or patch releases for `@primer/react`.

There are 2 approaches we can take:
1. (proposed in this ADR) Each experimental component is published as it's own npm package, example: `@primer/react-commentbox` (versioned individually)
2. Publish all experimental components together in a package `@primer/react-candidates` (Pro: One package to install and import from, less dependency management overhead. Con: A breaking change in one component is a breaking/major release for the whole package, can't update one component without making changes for the rest as well. Relevant because we expect more breaking changes in experimental components)

For the code repository, there are 2 approaches we can take:
1. Keep experimental components in [github.com/primer/react](http://github.com/primer/react), publish two or multiple packages from one repository. (Less overhead, one place where all the code lives)
2. (proposed in this ADR) Keep all experimental packages in one new repository [github.com/primer/react-candidates](http://github.com/primer/react-candidates). (this repository can have different conventions and processes with shared ownership between product teams and primer)

Reply:
I think I haven't done a good job at distinguishing the strategy for npm package and code repository. So here we go:

Right now, we have 1 code repository [github.com/primer/react](http://github.com/primer/react) and 1 npm package: [`@primer/react`](https://www.npmjs.com/@primer/react).

For the npm package, the proposal is publish experimental components outside of the npm package for `@primer/react` to keep the confidence/trust that there are no breaking changes in minor or patch releases for `@primer/react`.

There are 2 approaches we can take:
1. (proposed in this ADR) Each experimental component is published as it's own npm package, example: `@primer/react-commentbox` (versioned individually)
2. Publish all experimental components together in a package `@primer/react-candidates` (Pro: One package to install and import from, less dependency management overhead. Con: A breaking change in one component is a breaking/major release for the whole package, can't update one component without making changes for the rest as well. Relevant because we expect more breaking changes in experimental components)

For the code repository, there are 2 approaches we can take:
1. Keep experimental components in [github.com/primer/react](http://github.com/primer/react), publish two or multiple packages from one repository. (Less overhead, one place where all the code lives)
2. (proposed in this ADR) Keep all experimental packages in one new repository [github.com/primer/react-candidates](http://github.com/primer/react-candidates). (this repository can have different conventions and processes with shared ownership between product teams and primer)

Reply:
Added more clarity to `Decision` and `Other options considered` based on the comment above ^

Reply:
As far as whether to put it in a new vs same repo: It seems like a major consideration here is whether we want these experimental components to be open-source. Are we trying to experiment publicly, or would we rather keep the experimenting internal and only publicize polished work? If we want closed-source candidates, then we need a new, private, repo.

Reply:
> Are we trying to experiment publicly, or would we rather keep the experimenting internal and only publicize polished work?

I don't think we have considered that as a factor because we don't mind putting unpolished work in primer/react repo (open source)

There are times when a component might give a way a secret feature we are working on, but that's an exception not a rule